src/Capgemini.PowerApps.PackageDeployerTemplate/Adapters/CrmServiceAdapter.cs

```diff
@@ -313,6 +315,72 @@ public TResponse Execute<TResponse>(OrganizationRequest request, string username
             return (TResponse)this.Execute(request);
         }
 
+        /// <inheritdoc/>
+        public bool HasStartedSolutionHistoryRecords(ILogger logger)
+        {
+            var retryPolicy = Policy
+                .HandleResult<bool>(hasActiveRecords => !hasActiveRecords)
+                .WaitAndRetryForever(
+                    sleepDurationProvider: retryAttempt => TimeSpan.FromSeconds(20),
+                    onRetry: (result, timeSpan) =>
+                    {
+                        logger.LogInformation($"Active records still present within Solution History.  Waiting for {timeSpan.TotalSeconds} seconds before retrying...");
+                    });
+
+            var result = retryPolicy.Execute(() =>
+            {
+                var entityCollection = this.crmSvc.RetrieveMultiple(new QueryExpression()
+                {
+                    EntityName = Constants.SolutionHistory.LogicalName,
+                    ColumnSet = new ColumnSet(
+                    Constants.SolutionHistory.Fields.SolutionHistoryId,
+                    Constants.SolutionHistory.Fields.Name,
+                    Constants.SolutionHistory.Fields.Status),
+                    Criteria =
+                    {
+                        Conditions =
+                        {
+                            new ConditionExpression(Constants.SolutionHistory.Fields.Status, ConditionOperator.Equal, Constants.SolutionHistory.Statuses.Started),
+                        },
+                    },
+                });
+                return entityCollection.TotalRecordCount > 0;
+            });
+
+            return result;
+        }
+
+        /// <inheritdoc/>
+        public ExecuteMultipleResponse ExecuteMultipleSolutionHistoryOperation(IEnumerable<OrganizationRequest> requests, string username, ILogger logger, int? timeout = null)
+        {
+            ExecuteMultipleResponse executeMultipleRes = null;
+
+            var retryPolicy = Policy
+                .Handle<SolutionHistoryOperationException>()
+                .WaitAndRetryForever(
+                    sleepDurationProvider: retryAttempt => TimeSpan.FromSeconds(10),
+                    onRetry: (ex, timeSpan) =>
+                    {
+                        this.HasStartedSolutionHistoryRecords(logger);
+                    });
+
+            retryPolicy.Execute(() =>
```

Comment:
This looks like it will retry all requests, not just the ones that have failed. Can we change it so that successful requests are not retried?

Reply:
@ewingjm added a lambda expression to identify failed requests based on the returned responses.